Check for nil value in before_test and modified check for calling after_suite in before_test

mbround18 opened the pull request:
In response to #248 in the file `lib/minitest/reporters/base_reporter.rb` after the check for the nil the code was in place `after_suite` was still repetitively called and by adding the text check for `'Minitest::Result'` this will now call `after_suite` at the appropriate time. I tested on Minitest 5.11.1 using the SpecReporter, RubyMineReporter, DefaultReporter, and a custom reporter I am using. I followed that up by testing back on 5.10.3 to verify that it works on the older version as well.

os97673 commented on lib/minitest/reporters/base_reporter.rb:
```diff
@@ -1,3 +1,4 @@
+require 'pp'
```
redundant require?

os97673 commented on lib/minitest/reporters/base_reporter.rb:
```diff
@@ -14,11 +15,9 @@ def add_defaults(defaults)
-          after_suite(last_test.class) if last_test
-          before_suite(test.class)
-        end
+      	last_test = tests.last
```
it looks like indentation is broken

mbround18 replied:
Fixed indentation

os97673 commented:
to be able to review the changes it would be helpful if you'd describe the cause of the problem w/o it the diff looks like a collection of some random changes which happen to produce desired outcome :(

sandstrom commented:
@os97673 I think what's happened here is that in [`00433fc0`](https://github.com/seattlerb/minitest/commit/00433fc0a4fdd0e6b302aace633384ba13122376#diff-efcebb0b0e7d5c021792f400fd891d9fR930) a change was introduced where `run` would return an instance of `Result`. 

Because of this change the conditional `if last_test.class != test.class` (in this diff/PR) will never be true, so the `before_suite` callback is called before every test.

Does that make sense?

*(I agree with you that `pp` shouldn't be there!)*

mbround18 commented:
I added a commit which removes the redundant require. (I was using that for debugging and didn't mean to commit it) I also fixed the indentation as per comment. The reason for the change is the logic behind the initial code calls `before_suite` repetitively so in other reporters that do something like puts or streams the test class name to IO causes the before to be called every time. Next I made a nil check because when I was running the code locally tests.last does not initially exist because it is an empty array to begin with. Following that I noticed that `after_suite` was still being called after every test; So with a little debugging using `pp` I looked into what was happening and on the last test in the suite the object for `Minitest::Result` had not yet become an invocation of the actual class which is why I added the simple string check on the final test so it will than call `after_suite`.

os97673 commented:
Thank you for the explanation.  Need to think about it.  Still have some questions: e.g. why we call before_suite() and after_suite() in different calls? or why this code works for older versions of minitest.

mbround18 commented:
@os97673 I looked into this a bit and when you are on minitest version `>= 5.10.3` and modify master in the `BaseReporter` to include the following code and run a test suite with a couple test in it (in the return I added comments note that the comments wouldnt be there if you ran it yourself :P ):
```ruby
      # called by our own before hooks
      def before_test(test)
        last_test = tests.last
        pp last_test.class
        pp test.class
        if last_test.class != test.class
          after_suite(last_test.class) if last_test
          before_suite(test.class)
        end
      end
```
It will return with:
```ruby
NilClass # => run on TestSuite#test_001
TestSuite  # => run on TestSuite#test_001
TestSuite  # => run on TestSuite#test_002
TestSuite  # => run on TestSuite#test_002
TestSuite  # => run on TestSuite#test_003
TestSuite  # => run on TestSuite#test_003
TestSuite  # => run on TestSuite#test_004
TestSuite  # => run on TestSuite#test_004
```

However when you update to minitest `<= 5.11.0` the return will be:
```ruby 
NilClass # => run on TestSuite#test_001
TestSuite # => run on TestSuite#test_001
Minitest::Result # => run on TestSuite#test_002
TestSuite # => run on TestSuite#test_002
Minitest::Result # => run on TestSuite#test_003
TestSuite # => run on TestSuite#test_003
Minitest::Result # => run on TestSuite#test_004
TestSuite # => run on TestSuite#test_004
```


The culprit for the the issue lies within the `MiniTest::Test`  in the commit https://github.com/seattlerb/minitest/commit/00433fc0a4fdd0e6b302aace633384ba13122376#diff-444fa8babd8886c012733b064d709db5 when they changed how they return in the method `run` from being an instance of itself (itself being the test class as seen in the first result) to the return being an instance of the new class called `Result` see below, the change occured on line 104 of the `test.rb` file in the minitest code repository: 
```ruby
-        self # per contract
+       Result.from self # per contract
```

Side note I have a bit of a code change which leans up my code a bit so the code base wouldnt be dependent on minitest maintaining the name `MiniTest::Result` see my following commit.

sandstrom commented on lib/minitest/reporters/base_reporter.rb:
```diff
@@ -15,9 +15,10 @@ def add_defaults(defaults)
-          after_suite(last_test.class) if last_test
+        if last_test.nil?
           before_suite(test.class)
+        elsif is_a?(last_test.class) == String
```
Minor comment, but how about `elsif last_test.class.is_a?(String)` here?

sandstrom commented:
@os97673 I think @mbround18 explained it well, but otherwise let me know and I'll try to explain to the best of my ability.

mbround18 commented:
I worked with @aarontc on looking to the issue and him and I worked out a better solution to this issue, I am closing this pull request and am going to open a new one with this one as a reference.
